core/src/apps/cardano/sign_tx.py

```diff
@@ -139,9 +179,51 @@ def _build_witnesses(
     inputs: List[CardanoTxInputType],
     tx_body_hash: bytes,
     protocol_magic: int,
+) -> Dict:
+    shelley_witnesses = _build_shelley_witnesses(keychain, inputs, tx_body_hash)
+    byron_witnesses = _build_byron_witnesses(
+        keychain, inputs, tx_body_hash, protocol_magic
+    )
+
+    witnesses = {}
+    if len(shelley_witnesses) > 0:
+        witnesses[0] = shelley_witnesses
+    if len(byron_witnesses) > 0:
+        witnesses[2] = byron_witnesses
```

Comment:
Why is `[1]` left out? Is that by "definition" that Shelley witness is in position "0" and Byron in "2"?

Reply:
Yes, you can take a look at [shelley.cddl](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl#L192). Should I add a comment here or make the `0` and `2` into a named constant?

Reply:
I think comment above `witnesses = {}` would be just fine. And include the link please.

Reply:
Added [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/4d0a2c741e8365b7ffec319ac32b8d115d585b94). However, I am not sure about including the link, since it might change at anytime.

Reply:
You can add `https://github.com/input-output-hk/cardano-ledger-specs/blob/c453fa6fce99ed3cb09d129c829ddbf0980a642c/shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl` (you get that by pressing `y` in browser), this link won't change. However the link is ridiculously large then. Maybe let's just add vague "in shelley.cddl in cardano-ledger-specs"? 

Or you know what - what would be even better if you would create a README file in the app's root similar as Ripple has. And put some interesting and important links there. It can help a lot some time in future.

Reply:
I've noticed the README files and I would like to create one, but I would prefer to do that after also the third PR is done (or as a part of).

Reply:
Link removed [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/bc44b901866e6ea1f734172e09da47567b0fd4a0).

Reply:
Sounds alright!